Source/Shared/BPConfiguration.m

```diff
@@ -30,58 +30,65 @@
         "The number of retries on any failures (app crash/test failure)."},
-    {'h', "help",     no_argument, NULL, BP_VALUE, NULL,
-     "This help."},
-    {'i', "include", required_argument, NULL, BP_LIST, "testCasesToRun",
-     "Include a testcase in the set of tests to run."},
+    {'i', "include", optional_argument, NULL, BP_LIST, "testCasesToRun",
+        "Include a testcase in the set of tests to run."},
     {'j', "junit-output", no_argument, "Off", BP_VALUE | BP_BOOL, "junitOutput",
-     "Print results in JUnit format."},
+        "Print results in JUnit format."},
     {'l', "list-tests", no_argument, NULL, BP_VALUE, "listTestsOnly",
-     "Only list tests in bundle"},
-    {'n', "num-sims", required_argument, "4", BP_VALUE, "numSims",
-     "Number of simulators to run in parallel. (bluepill only)"},
-    {'o', "output-dir", required_argument, NULL, BP_VALUE | BP_PATH, "outputDirectory",
-     "Directory where to put output log files (bluepill only)."},
+        "Only list tests in bundle"},
+    {'n', "num-sims", optional_argument, "4", BP_VALUE, "numSims",
+        "Number of simulators to run in parallel. (bluepill only)"},
+    {'r', "runtime",  optional_argument, BP_DEFAULT_RUNTIME, BP_VALUE, "runtime",
+        "What runtime to use."},
+    {'t', "test",     optional_argument, NULL, BP_VALUE | BP_PATH, "testBundlePath",
+        "The path to the test bundle to execute (your .xctest)."},
+    {'x', "exclude", optional_argument, NULL, BP_LIST, "testCasesToSkip",
+        "Exclude a testcase in the set of tests to run."},
+    {'X', "xcode-path", optional_argument, NULL, BP_VALUE | BP_PATH, "xcodePath",
+        "Path to xcode."},
+
+    // options with no argument
+    {'H', "headless", no_argument, "Off", BP_VALUE | BP_BOOL , "headlessMode",
+        "Run in headless mode (no GUI)."},
+    {'J', "json-output", no_argument, "Off", BP_VALUE | BP_BOOL, "jsonOutput",
+        "Print test timing information in JSON format."},
+    {'h', "help",     no_argument, NULL, BP_VALUE, NULL,
+        "This help."},
     {'p', "plain-output", no_argument, "Off", BP_VALUE | BP_BOOL, "plainOutput",
-     "Print results in plain text."},
+        "Print results in plain text."},
     {'q', "quiet", no_argument, "Off", BP_VALUE | BP_BOOL, "quiet",
         "Turn off all output except fatal errors."},
-    {'r', "runtime",  required_argument, BP_DEFAULT_RUNTIME, BP_VALUE, "runtime",
-     "What runtime to use."},
-    {'s', "scheme-path", required_argument, NULL, BP_VALUE | BP_PATH, "schemePath",
-        "The scheme to run tests."},
-    {'t', "test",     required_argument, NULL, BP_VALUE | BP_PATH, "testBundlePath",
-     "The path to the test bundle to execute (your .xctest)."},
-    {'x', "exclude", required_argument, NULL, BP_LIST, "testCasesToSkip",
-     "Exclude a testcase in the set of tests to run."},
-    {'X', "xcode-path", optional_argument, NULL, BP_VALUE | BP_PATH, "xcodePath",
-        "Path to xcode."},
+
```

Comment:
`optional_argument` and `required_argument` are arguments _to the option_, not whether the option itself is optional or required. For example, '-n' takes a `required_argument` because you *must* specify how many simulators to run. It doesn't mean that `-n` itself is required.

Also, it seems `getopt_long()` on macOS has a bug where `optional_argument` doesn't work properly so I was thinking of just making _everything_ that takes an argument a `required_argument`.

We need a test in `build.sh` that runs `bluepill` from the command line to test argument processing because no test caught this problem.